Add support for footnotes

BrewingWeasel opened the pull request:
This PR adds support for footnotes.
Due to the way footnotes work in Djot, I had to restructure parts of the existing code to add support for parsing footnote definition blocks and incrementing the footnote number for each reference in the generated html.

BrewingWeasel committed:
add support for footnotes

BrewingWeasel committed:
fix ordering of attributes to pass test

BrewingWeasel committed:
fix not including empty footnotes

BrewingWeasel committed:
fix using footnotes inside of other footnotes

BrewingWeasel committed:
update CHANGELOG and README

lpil committed:
Update CI version

lpil commented on src/jot.gleam:
```diff
@@ -99,71 +107,181 @@ fn drop_spaces(in: Chars) -> Chars {
+  )
+}
+
+fn parse_block_indent_unchecked(
```
What do these new functions do? With the "unknown handler" arguments. I'm not sure I understand the design yet. Some documentation would be useful, given they are more complex than before.

What does "unchecked" mean also?

lpil reviewed (commented):
Great work! Thank you!

I've left a question inline. I think the code would benefit from some documentation comments as it has got more complex- there's a good deal more abstraction being applied after this patch than the previous design had. This can be OK, but it makes the code harder to understand and modify in future, so we need to make it as clear as possible.

Thank you.

P.S. You sentence mining project looks really cool! Maybe one day I'll figure you how to use it for learning Irish Gaeilge :)

BrewingWeasel committed:
slightly refactor and add documentation

BrewingWeasel committed:
fix not properly indenting codeblock directly following footnote definition

BrewingWeasel commented:
Thanks for the review!

I was able to factor out some of the abstraction because it turns out you can nest definitions (which I thought I had already checked but I guess not 🤷).
I also added some more comments and renamed some functions, so hopefully everything is more clear.

Additionally, I noticed a niche case that wouldn't be handled correctly, so I fixed that and added a test for it.

Also, best of luck with your Irish Gaeilge learning!

lpil reviewed (approved):
Thank you!! Fab work!
